benchmarks/GraphBLAS-sharp.Benchmarks/Algorithms/PageRank.fs

```diff
@@ -0,0 +1,132 @@
+    let mutable matrixHost = Unchecked.defaultof<_>
+
+    let accuracy = 0.00000001f
+
+    member val Result = Unchecked.defaultof<ClVector<float32>> with get,set
+
+    [<ParamsSource("AvailableContexts")>]
+    member val OclContextInfo = Unchecked.defaultof<Utils.BenchmarkContext * int> with get, set
+
+    [<ParamsSource("InputMatrixProvider")>]
+    member val InputMatrixReader = Unchecked.defaultof<MtxReader> with get, set
+
+    member this.OclContext = (fst this.OclContextInfo).ClContext
+    member this.WorkGroupSize = snd this.OclContextInfo
+
+    member this.Processor =
+        let p = (fst this.OclContextInfo).Queue
+        p.Error.Add(fun e -> failwithf "%A" e)
+        p
+
+    static member AvailableContexts = Utils.availableContexts
+
+    static member InputMatrixProviderBuilder pathToConfig =
+        let datasetFolder = ""
+        pathToConfig
+        |> Utils.getMatricesFilenames
+        |> Seq.map
+            (fun matrixFilename ->
+                printfn "%A" matrixFilename
+
+                match Path.GetExtension matrixFilename with
+                | ".mtx" -> MtxReader(Utils.getFullPathToMatrix datasetFolder matrixFilename)
+                | _ -> failwith "Unsupported matrix format")
+
+    member this.FunToBenchmark =
+        match funToBenchmark with
+        | None ->
+            let x = buildFunToBenchmark this.OclContext this.WorkGroupSize
+            funToBenchmark <- Some x
+            x
+        | Some x -> x
+
+    member this.PageRank() =
+        this.Result <- this.FunToBenchmark this.Processor matrixPrepared accuracy
+
+    member this.ClearInputMatrix() =
+        matrix.Dispose this.Processor
+
+    member this.ClearPreparedMatrix() =
+        matrixPrepared.Dispose this.Processor
+
+    member this.ClearResult() = this.Result.Dispose this.Processor
+
+    member this.ReadMatrix() =
+        let converter =
+            match this.InputMatrixReader.Field with
+            | Pattern -> binaryConverter
+            | _ -> converter
+
+        matrixHost <- this.InputMatrixReader.ReadMatrix converter
+
+    member this.LoadMatrixToGPU() =
+        matrix <- buildMatrix this.OclContext matrixHost
+
+    member this.PrepareMatrix() =
+        matrixPrepared <- Algorithms.PageRank.prepareMatrix this.OclContext this.WorkGroupSize this.Processor matrix
+
+    abstract member GlobalSetup : unit -> unit
+
+    abstract member IterationCleanup : unit -> unit
+
+    abstract member GlobalCleanup : unit -> unit
+
+    abstract member Benchmark : unit -> unit
+
+type PageRankWithoutTransferBenchmarkFloat32() =
+
+    inherit Benchmarks(
+        Algorithms.PageRank.run,
+        float32,
+        (fun _ -> float32 <| Utils.nextInt (System.Random())),
+        (fun context matrix -> ClMatrix.CSR <| matrix.ToCSR.ToDevice context))
+
+    static member InputMatrixProvider =
+        Benchmarks.InputMatrixProviderBuilder "BFSBenchmarks.txt"
+
+    [<GlobalSetup>]
+    override this.GlobalSetup() =
```

Comment:
It is a good idea to block after preparotory methods in case some of them are not blocking by itself, but blocking in the `[<Benchmark>]` is still necessary to wait for all kernels to execute.